compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -4187,24 +4194,48 @@ object Types {
       // To do this, we need first test cases for that situation.
 
     /** Instantiate variable with given type */
-    def instantiateWith(tp: Type)(implicit ctx: Context): Type = {
+    def instantiateWith(tp: Type)(implicit ctx: Context): Unit =
       assert(tp ne this, s"self instantiation of ${tp.show}, constraint = ${ctx.typerState.constraint.show}")
       typr.println(s"instantiating ${this.show} with ${tp.show}")
-      if ((ctx.typerState eq owningState.get) && !ctx.typeComparer.subtypeCheckInProgress)
+      if (ctx.typerState eq owningState.get) && !ctx.typeComparer.subtypeCheckInProgress then
         inst = tp
       ctx.typerState.constraint = ctx.typerState.constraint.replace(origin, tp)
-      tp
-    }
+
+    /** The type this type variable should be instantiated with.
+     *  This is the type-comparer's instance type with the added tweak that
+     *  captured for more deeply nested symbols are avoided.
+     */
+    def instanceType(fromBelow: Boolean)(using Context): Type =
+      avoidCaptures(ctx.typeComparer.instanceType(origin, fromBelow))
 
     /** Instantiate variable from the constraints over its `origin`.
-     *  If `fromBelow` is true, the variable is instantiated to the lub
-     *  of its lower bounds in the current constraint; otherwise it is
-     *  instantiated to the glb of its upper bounds. However, a lower bound
-     *  instantiation can be a singleton type only if the upper bound
-     *  is also a singleton type.
+     *  If `fromBelow` is true, the variable's instance type is computed from
+     *  the lub of its lower bounds in the current constraint; otherwise it is
+     *  computed from the glb of its upper bounds.
+     *  After that, there are some further transformations and checks:
+     *
+     *   - Singletons and union types are sometimes widened according to
+     *     `ConstraintHandling.widenInferred`.
+     *   - References to more neeply nested symbols are avoided
+     *     according to `avoidCaptures`
+     *   - It is checked that the resulting type still fits within
+     *     bounds. If not, a `NoInstance` exception is thrown,
+     *     but only after instantiating the variable anyway because
+     *     this avoids follow-on errors.
```

Comment:
I'm not convinced this is needed:
- `avoidCaptures` already checks that the approximation is within bound and throw an exception if it isn't
- `widenInferred` will not do any widening if that would be out of bounds

The only problematic piece is `approximation` in `ConstraintHandling` which will replace recursive occurences of the type variable being instantiated by `TypeBounds.empty`, which can indeed be unsound. Here's an alternative PR which tries to fix that specific problem: https://github.com/lampepfl/dotty/pull/9012